### PR TITLE
Compress non-neighbor stops.

### PR DESCRIPTION
**Pull request by siarheishut**

### CODE REVIEW

File: src/coords_converter.cpp [hunk 1]
```diff
@@ -1,49 +1,106 @@
...
+  if (stop_count == 1) {
+    vector<pair<string_view, double>> result;
+    for (auto stop : layers.front()) {
+      result.emplace_back(stop, from);
+    }
+    return result;
   }
 
+  vector<pair<string_view, double>> result;
+  double step = (to - from) / (static_cast<double>(layers.size()) - 1);
```
Comment (foxfrikses) [inline src/coords_converter.cpp:98]: `(static_cast<double>(layers.size())` == `stop_count`

Comment (foxfrikses) [reply src/coords_converter.cpp:98]:
```
double step = (stop_count == 1) ? 0.0 :  (to - from) / (stop_count - 1);
```
And you can completely delete remove `if (stop_count == 0)` and `if (stop_count == 1)` cases.

File: src/coords_converter.cpp [hunk 2]
```diff
@@ -1,49 +1,106 @@
...
+  for (int i = 0; i < layers.size(); ++i) {
+    for (auto stop : layers[i]) {
+      result.emplace_back(stop, from + step * i);
```
Comment (foxfrikses) [inline src/coords_converter.cpp:101]: transform? =)

File: src/coords_converter.h [hunk 3]
```diff
@@ -2,18 +2,35 @@
...
+#include "map_renderer_utils.h"
 #include "request_types.h"
 #include "sphere.h"
 
-namespace rm {
-std::unordered_map<std::string_view, svg::Point> ConvertCoords(
-    const std::map<std::string_view, rm::sphere::Coords> &stops,
-    const rm::RenderingSettings &settings);
+namespace rm::coords_converter {
+using namespace std;
```
Comment (foxfrikses) [inline src/coords_converter.h:16]: No using namespace in .h

File: src/coords_converter.h [hunk 4]
```diff
@@ -2,18 +2,35 @@
...
+
+enum class SortMode { kByLatitude, kByLongitude };
+
+auto SortStops(const renderer_utils::Stops &stops, SortMode mode)
+-> vector<string_view>;
+
+auto AdjacentStops(const renderer_utils::Buses &buses)
+-> vector<pair<string_view, string_view>>;
+
+auto CompressNonNeighbors(
```
Comment (foxfrikses) [inline src/coords_converter.h:26]: Maybe then `CompressNonadjacent` (to be consistent with `AdjacentStops` above?

File: src/coords_converter.h [hunk 5]
```diff
@@ -2,18 +2,35 @@
...
+    const vector<string_view> &stops,
+    const vector<pair<string_view, string_view>> &neighbors)
+-> vector<vector<string_view>>;
```
Comment (foxfrikses) [inline src/coords_converter.h:29]: Let's indeed use standard way to define return type (without auto ->)

File: src/map_renderer.cpp [hunk 6]
```diff
@@ -51,6 +51,46 @@ svg::Text BusNameText(std::string name, svg::Point point,
                        .SetData(std::move(name))
                        .SetFillColor(color));
 }
+
+struct CoordsConverterSettings {
+  double width, height;
+  double padding;
+};
+
+auto CombineCoords(std::vector<std::pair<std::string_view, double>> x_coords,
```
Comment (foxfrikses) [inline src/map_renderer.cpp:60]: why not `const&`

File: src/map_renderer.cpp [hunk 7]
```diff
@@ -113,8 +157,7 @@ MapRenderer::MapRenderer(
 }
 
 void MapRenderer::AddBusLinesLayout(
-    const renderer_utils::Buses &buses,
```
Comment (foxfrikses) [inline src/map_renderer.cpp:116]: why here

File: src/map_renderer.cpp [hunk 8]
```diff
@@ -139,8 +182,7 @@ void MapRenderer::AddBusLinesLayout(
 }
 
 void MapRenderer::AddBusLabelsLayout(
-    const renderer_utils::Buses &buses,
```
Comment (foxfrikses) [inline src/map_renderer.cpp:142]: why here

File: src/map_renderer.h [hunk 9]
```diff
@@ -18,8 +18,7 @@ namespace rm {
 class MapRenderer {
  public:
   static std::unique_ptr<MapRenderer> Create(
-      const renderer_utils::Buses &buses,
```
Comment (foxfrikses) [inline src/map_renderer.h:21]: why here

File: src/map_renderer.h [hunk 10]
```diff
@@ -29,14 +28,12 @@ class MapRenderer {
               const RenderingSettings &settings);
 
   void AddBusLinesLayout(
-      const renderer_utils::Buses &buses,
```
Comment (foxfrikses) [inline src/map_renderer.h:32]: why here

File: src/map_renderer.h [hunk 11]
```diff
@@ -29,14 +28,12 @@ class MapRenderer {
...
-      const renderer_utils::Stops &stops,
+      const renderer_utils::Buses &buses, const renderer_utils::Stops &stops,
       const rm::RenderingSettings &settings,
       const renderer_utils::StopCoords &coords);
 
   void AddBusLabelsLayout(
-      const renderer_utils::Buses &buses,
```
Comment (foxfrikses) [inline src/map_renderer.h:38]: why here

File: src/request_processor.cpp [hunk 12]
```diff
@@ -8,21 +8,19 @@
...
 #include "bus_manager.h"
 #include "map_renderer.h"
-#include "map_renderer_utils.h"
 
 namespace {
 auto MapRendererParams(const std::vector<rm::PostRequest> &requests) {
   using namespace std;
   using namespace rm;
 
-  map<string_view, rm::renderer_utils::Route> buses;
```
Comment (foxfrikses) [inline src/request_processor.cpp:18]: why here

File: src/request_processor.cpp [hunk 13]
```diff
@@ -8,21 +8,19 @@
...
+  map<string_view, renderer_utils::Route> buses;
   map<string_view, sphere::Coords> stops;
   for (auto &request : requests) {
     if (auto bus = get_if<PostBusRequest>(&request)) {
-      buses.emplace(
```
Comment (foxfrikses) [inline src/request_processor.cpp:22]: why here

File: src/request_processor.cpp [hunk 14]
```diff
@@ -8,21 +8,19 @@
 
 #include "bus_manager.h"
 #include "map_renderer.h"
-#include "map_renderer_utils.h"
```
Comment (foxfrikses) [inline src/request_processor.cpp:11]: you use them here, don't delete them

File: src/map_renderer.cpp [hunk 15]
```diff
@@ -51,6 +51,46 @@ svg::Text BusNameText(std::string name, svg::Point point,
...
+
+auto TransformCoords(
+    const rm::renderer_utils::Buses &buses,
+    const rm::renderer_utils::Stops &stops,
+    const CoordsConverterSettings settings)
+-> std::unordered_map<std::string_view, svg::Point> {
+  using namespace rm::coords_converter;
+
+  auto stops_by_lon = SortStops(stops, SortMode::kByLongitude);
+  auto layers_by_lon = CompressNonNeighbors(stops_by_lon, AdjacentStops(buses));
```
Comment (foxfrikses) [inline src/map_renderer.cpp:81]: Why not store `auto neighbors = AdjacentStops(buses);` once.

File: src/map_renderer.cpp [hunk 16]
```diff
@@ -51,6 +51,46 @@ svg::Text BusNameText(std::string name, svg::Point point,
...
+                              settings.padding,
+                              settings.width - settings.padding);
+
+  auto stops_by_lat = SortStops(stops, SortMode::kByLatitude);
+  auto layers_by_lat = CompressNonNeighbors(stops_by_lat, AdjacentStops(buses));
+  auto y_coords = SpreadStops(layers_by_lat,
+                              settings.height - settings.padding,
+                              settings.padding);
+
+  return CombineCoords(std::move(x_coords), std::move(y_coords));
```
Comment (foxfrikses) [inline src/map_renderer.cpp:92]: I don't think you need to move these here

File: src/map_renderer.cpp [hunk 17]
```diff
@@ -82,7 +122,11 @@ MapRenderer::MapRenderer(
     const renderer_utils::Buses &buses,
     const renderer_utils::Stops &stops,
     const RenderingSettings &settings) {
-  auto stop_to_coords = ConvertCoords(stops, settings);
+  CoordsConverterSettings coords_settings = {
```
Comment (foxfrikses) [inline src/map_renderer.cpp:125]: Maybe you can store width, height and padding in a separate structure like `renderer_utils::Frame` inside RenderingSettings?

File: src/coords_converter.cpp [hunk 18]
```diff
@@ -1,49 +1,106 @@
...
 
-namespace rm {
-std::unordered_map<std::string_view, svg::Point> ConvertCoords(
-    const std::map<std::string_view, rm::sphere::Coords> &stops,
-    const rm::RenderingSettings &settings) {
-  std::unordered_map<std::string_view, svg::Point> result;
+namespace rm::coords_converter {
+auto SortStops(const renderer_utils::Stops &stops, SortMode mode)
+-> vector<string_view> {
+  auto f = [mode]() -> double (*)(rm::sphere::Coords) {
```
Comment (foxfrikses) [inline src/coords_converter.cpp:16]: maybe `coord` instead of `f`

File: src/coords_converter.cpp [hunk 19]
```diff
@@ -1,49 +1,106 @@
...
+  vector<pair<double, string_view>> sorted_stops;
+  transform(begin(stops), end(stops), back_inserter(sorted_stops),
+            [f](auto &item) {
+              return pair{f(item.second), item.first};
+            });
+  sort(begin(sorted_stops), end(sorted_stops));
+
+  vector<string_view> layers;
+  transform(begin(sorted_stops), end(sorted_stops), back_inserter(layers),
+            [](auto &item) {
```
Comment (foxfrikses) [inline src/coords_converter.cpp:31]: Why not in one like?

File: src/coords_converter.cpp [hunk 20]
```diff
@@ -1,49 +1,106 @@
...
-  std::sort(begin(stops_by_lon), end(stops_by_lon));
-
-  double x_step = (settings.width - 2 * settings.padding) / (stop_count - 1);
-  double y_step = (settings.height - 2 * settings.padding) / (stop_count - 1);
-  for (int i = 0; i < stops_by_lat.size(); ++i) {
-    result[stops_by_lat[i].second].y =
-        settings.height - settings.padding - y_step * i;
+
+  vector<vector<string_view>> res{{stops[0]}};
+  int last = max_next[0];
```
Comment (foxfrikses) [inline src/coords_converter.cpp:71]: You can make it 
```
vector<vector<string_view>> res;
int last = -1;
```
and remove `if (stops.empty()) return {};` in the begining of the function

File: tests/coords_converter_test.cpp [hunk 21]
```diff
@@ -0,0 +1,216 @@
...
+const std::pair<std::string_view, rm::sphere::Coords> kHighStreet =
+    {"High Street", {.latitude = 42.111111, .longitude = 19.291527}};
+const std::pair<std::string_view, rm::sphere::Coords> kRWStation =
+    {"RW station", {.latitude = 63.015232, .longitude = -35.621022}};
+const std::pair<std::string_view, rm::sphere::Coords> kClemens =
+    {"Clemens street", {.latitude = 49.102842, .longitude = 39.845290}};
+const std::pair<std::string_view, rm::sphere::Coords> kDostoevsky =
+    {"Dostoevsky street", {.latitude = 13.102921, .longitude = 12.432091}};
+
+//auto SortStops(const renderer_utils::Stops &stops, SortMode mode)
```
Comment (foxfrikses) [inline tests/coords_converter_test.cpp:23]: ?

File: tests/coords_converter_test.cpp [hunk 22]
```diff
@@ -0,0 +1,216 @@
...
+  };
+
+  std::vector<TestCase> test_cases{
+      TestCase{
+          .name = "Empty names",
+          .stops = {},
+          .want = {},
+      },
+      TestCase{
+          .name = "Stop count > 1",
```
Comment (foxfrikses) [inline tests/coords_converter_test.cpp:43]: what about `== 1`?

File: tests/coords_converter_test.cpp [hunk 23]
```diff
@@ -0,0 +1,216 @@
...
+    Stops stops;
+    std::vector<std::string_view> want;
+  };
+
+  std::vector<TestCase> test_cases{
+      TestCase{
+          .name = "Empty names",
+          .stops = {},
+          .want = {},
+      },
```
Comment (foxfrikses) [inline tests/coords_converter_test.cpp:70]: `== 1`

File: tests/coords_converter_test.cpp [hunk 24]
```diff
@@ -0,0 +1,216 @@
...
+    vector<pair<string_view, string_view>> neighbors;
+    vector<vector<string_view>> want;
+  };
+
+  vector<TestCase> test_cases{
+      TestCase{
+          .name = "Empty stops",
+          .want = {},
+      },
+      TestCase{
```
Comment (foxfrikses) [inline tests/coords_converter_test.cpp:136]: You can have:
1. the result contains, for example, one layer with 4 stops, one with 3, 2 with 2, 4 with 1.
2. empty stops
3. empty neighbors

File: tests/coords_converter_test.cpp [hunk 25]
```diff
@@ -0,0 +1,216 @@
...
+          .name = "One layer(Latitudes)",
+          .layers = {{"3", "1", "2"}},
+          .from = 300,
+          .to = 100,
+          .want = {{"3", 300}, {"1", 300}, {"2", 300}},
+      },
+      TestCase{
```
Comment (foxfrikses) [inline tests/coords_converter_test.cpp:189]: You don't distinguish longitudes and latitudes in this func

File: tests/coords_converter_test.cpp [hunk 26]
```diff
@@ -0,0 +1,216 @@
...
+          .to = 100,
+          .want = {{"3", 300}, {"5", 300}, {"1", 200}, {"2", 100}, {"4", 100}},
+      },
+      TestCase{
+          .name = "Layer count > 1(Longitude)",
+          .layers = {{"3", "5"}, {"1"}, {"2", "4"}},
+          .from = 100,
+          .to = 300,
+          .want = {{"3", 100}, {"5", 100}, {"1", 200}, {"2", 300}, {"4", 300}},
+      },
```
Comment (foxfrikses) [inline tests/coords_converter_test.cpp:209]: Add tests for `from > to`, and for `to > from`, and for `to == from`

File: tests/map_renderer_test.cpp [hunk 27]
```diff
@@ -344,39 +343,6 @@ TEST(TestMapRenderer, TestGetMap) {
                       STOP_NAME(Shop, 170, 170))
 
       },
-      TestCase{
-          .name = "Empty stop name",
```
Comment (foxfrikses) [inline tests/map_renderer_test.cpp:348]: What changes for empty stop name?

File: tests/map_renderer_test.cpp [hunk 28]
```diff
@@ -293,26 +292,26 @@ TEST(TestMapRenderer, TestGetMap) {
...
                       STOP_NAME(Airport, 242.1, 225.7)
-                      STOP_NAME(High Street, 218.067, 207.133)
-                      STOP_NAME(RW station, 194.033, 188.567)
+                      STOP_NAME(High Street, 206.05, 197.85)
+                      STOP_NAME(RW station, 206.05, 197.85)
                       STOP_NAME(Shop, 170, 170))
       },
       TestCase{
-          .name = "Route number is greater than palette size",
+          .name = "renderer_utils::Route number is greater than palette size",
```
Comment (foxfrikses) [inline tests/map_renderer_test.cpp:314]: I don't think it should be `renderer_utils::` here

File: tests/map_renderer_test.cpp [hunk 29]
```diff
@@ -9,7 +9,6 @@
 #include "gtest/gtest.h"
 #include "svg/common.h"
 
-#include "src/map_renderer_utils.h"
```
Comment (foxfrikses) [inline tests/map_renderer_test.cpp:12]: Why do you delete it if you use `renderer_utils::Route` in this file

File: tests/map_renderer_test.cpp [hunk 30]
```diff
@@ -173,7 +172,7 @@ TEST(TestMapRenderer, TestGetMap) {
 
   struct TestCase {
     std::string name;
-    std::map<std::string_view, rm::renderer_utils::Route> buses;
+    std::map<std::string_view, renderer_utils::Route> buses;
```
Comment (foxfrikses) [inline tests/map_renderer_test.cpp:175]: I don't think it's for this commit

File: src/coords_converter.cpp (hunk 1 above)
Comment (siarheishut) [reply src/coords_converter.cpp:98]: Done.

File: src/coords_converter.cpp (hunk 2 above)
Comment (siarheishut) [reply src/coords_converter.cpp:101]: =)

File: src/coords_converter.h (hunk 3 above)
Comment (siarheishut) [reply src/coords_converter.h:16]: Removed.

File: src/coords_converter.h (hunk 4 above)
Comment (siarheishut) [reply src/coords_converter.h:26]: Done.

File: src/coords_converter.h (hunk 5 above)
Comment (siarheishut) [reply src/coords_converter.h:29]: Done.

File: src/map_renderer.cpp (hunk 6 above)
Comment (siarheishut) [reply src/map_renderer.cpp:60]: Done.

File: src/coords_converter.cpp [hunk 31]
```diff
@@ -1,52 +1,93 @@
...
-  }
+using namespace std;
+
+namespace rm::coords_converter {
+std::vector<std::string_view>
+SortStops(const renderer_utils::Stops &stops, SortMode mode) {
+  auto coord = [mode]() -> double (*)(rm::sphere::Coords) {
+    if (mode == SortMode::kByLatitude)
+      return [](auto coords) { return coords.latitude; };
+    return [](auto coords) { return coords.longitude; };
```
Comment (foxfrikses) [inline src/coords_converter.cpp:21]: I think it'd be easier to just write if in transform lambda

File: tests/coords_converter_test.cpp [hunk 32]
```diff
@@ -0,0 +1,230 @@
...
+      TestCase{
+          .name = "to == from",
+          .layers = {{"3", "1", "2"}},
+          .from = 100,
+          .to = 100,
+          .want = {{"3", 100}, {"1", 100}, {"2", 100}},
+      },
+      TestCase{
+          .name = "Layer count > 1(Latitudes)",
```
Comment (foxfrikses) [inline tests/coords_converter_test.cpp:211]: I think you need to remove `"Layer count > 1(Latitudes)"` and `"Layer count > 1(Longitude)"` and modify `to == from`, `to > from` and `from > to` in order for them to process multiple layers.
And I think you should have `to == from`, `to > from` and `from > to` tests for 1 layer as well as for multiple. 
7 tests overall.

File: src/coords_converter.cpp (hunk 31 above)
Comment (siarheishut) [reply src/coords_converter.cpp]: Done.

File: tests/coords_converter_test.cpp (hunk 32 above)
Comment (siarheishut) [reply tests/coords_converter_test.cpp]: Done.

**Review by foxfrikses — APPROVED** [review]
LGTM

**Pull request merged**